KubernetesPodOperator doc (#175)
astronomer/docs: cloud/kubernetespodoperator.md

---
sidebar_label: 'KubernetesPodOperator'

> title: "Run the KubernetesPodOperator on Astronomer Cloud"
> id: kubernetespodoperator
> ---
> 
> ## Overview
> 
> This guide provides steps for configuring and running the KubernetesPodOperator on DAGs deployed to Astronomer Cloud.

Reviewer: ```suggestion
title: "Run the KubernetesPodOperator on Astro"
id: kubernetespodoperator
---

## Overview

This guide provides steps for configuring and running the KubernetesPodOperator on DAGs deployed to Astro.
```

> The [KubernetesPodOperator](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html) is one of Apache Airflow's most powerful operators. Similar to the Kubernetes Executor, this operator talks to the Kubernetes API to dynamically launch a pod in Kubernetes for each task that needs to run and terminates each pod once the task is completed. This results in an isolated, containerized execution environment for each task that is separate from tasks otherwise being executed by Celery workers. The KubernetesPodOperator enables you to:

Reviewer: ```suggestion
The [KubernetesPodOperator](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html) is one of Apache Airflow's most powerful operators. Similar to the Kubernetes Executor, this operator talks to the Kubernetes API to dynamically launch a Pod in Kubernetes for each task that needs to run and terminates each Pod once the task is completed. This results in an isolated, containerized execution environment for each task that is separate from tasks otherwise being executed by Celery workers. The KubernetesPodOperator enables you to:
```

- Execute a custom Docker image per task with Python packages and dependencies that would otherwise conflict with the rest of your Deployment's dependencies. This includes Docker images in a private registry or repository.

> - Specify CPU and Memory as task-level limits or minimums to optimize costs and performance.

Reviewer: ```suggestion
- Specify CPU and Memory as task-level limits or minimums to optimize for cost and performance.
```

- Write task logic in a language other than Python. This gives you flexibility and can enable new use cases across teams.
- Scale task growth horizontally in a way that is cost-effective, dynamic, and minimally dependent on Worker resources.

> - Access images hosted in private Docker repositories.

Reviewer: ```suggestion
```

Redundant w above

- Set Kubernetes-native configurations in a YAML file, including volumes, secrets, and affinities.

> On Astronomer Cloud, the Kubernetes infrastructure required to run the KubernetesPodOperator is built into every Cluster in the Data Plane and is managed by Astronomer.

Reviewer: ```suggestion
On Astro, the Kubernetes infrastructure required to run the KubernetesPodOperator is built into every Cluster in the Data Plane and is managed by Astronomer.
```

## Prerequisites

To use the KubernetesPodOperator, you need:

> - An [Astronomer project](create-project.md).
> - An Astronomer [Deployment](configure-deployment.md).

Reviewer: ```suggestion
- An [Astro project](create-project.md).
- An Astro [Deployment](configure-deployment.md).
```

## Set Up the KubernetesPodOperator

To use the KubernetesPodOperator in a DAG, add the following import statements and instantiation to your DAG file:

```python
from airflow.contrib.operators.kubernetes_pod_operator import kubernetes_pod_operator

# Pulls environment information from Astronomer Cloud
from airflow import configuration as conf
...

namespace = conf.get('kubernetes', 'NAMESPACE')
k = kubernetes_pod_operator.KubernetesPodOperator(
    namespace=namespace,
    image="<your-docker-image>",
    cmds=["<commands-for-image>"],
    arguments=["<arguments-for-image>"],
    labels={"<pod-label>": "<label-name>"},
    name="<pod-name>",
    is_delete_operator_pod=True,
    in_cluster=True,
    task_id="<task-name>",
    get_logs=True)
```

For each instantiation of the KubernetesPodOperator, you must specify the following values:

> - `namespace = conf.get('kubernetes', 'NAMESPACE')`: Every Astronomer Deployment runs on its own Kubernetes namespace within a Cluster. Information about this namespace can be programmatically imported as long as you set this variable.

Reviewer: ```suggestion
- `namespace = conf.get('kubernetes', 'NAMESPACE')`: Every Astro Deployment runs on its own Kubernetes namespace within a Cluster. Information about this namespace can be programmatically imported as long as you set this variable.
```

- `image`: This is the Docker image that the operator will use to run its defined task, commands, and arguments. The value you specify is assumed to be an image tag that's publicly available on [Docker Hub](https://hub.docker.com/). To pull an image from a private registry, read [Pull Images from a Private Registry](kubernetespodoperator.md#pull-images-from-a-private-registry).

> - `in_cluster=True`: When this value is set, your task will run within the Cluster from which it's instantiated on Astronomer. This ensures that the Kubernetes pod running your task has the correct permissions within the Cluster.

Reviewer: ```suggestion
- `in_cluster=True`: When this value is set, your task will run within the Cluster from which it's instantiated on Astro. This ensures that the Kubernetes Pod running your task has the correct permissions within the Cluster.
```

- `is_delete_operator_pod=True`: This setting ensures that once a KubernetesPodOperator task is complete, the Kubernetes Pod that ran that task is terminated. This ensures that there are no unused pods in your Cluster taking up resources.

> This is the minimum configuration required to run tasks with the KubernetesPodOperator on Astronomer Cloud. To further customize the way your tasks are run, see the topics below.

Reviewer: ```suggestion
This is the minimum configuration required to run tasks with the KubernetesPodOperator on Astro. To further customize the way your tasks are run, see the topics below.
```

        is_delete_operator_pod=True,
        get_logs=True)
```

## Run Images from a Private Registry

By default, the KubernetesPodOperator expects to pull a Docker image that's hosted publicly on Docker Hub. If you want to execute a Docker image that's hosted in a private registry, complete the setup below.

### Prerequisites

To complete this setup, you need:

- A private registry that hosts your images.
- kubectl.
- An Astronomer Cloud project.
- A Deployment hosting your project.

### Step 1: Create a Kubernetes Secret

To run Docker images from a private registry on Astronomer Cloud, you first need to create a Kubernetes secret that contains credentials to your registry. Astronomer will then inject that secret into your Deployment's namespace, which will give your tasks access to Docker images within your registry. To do this, complete one of the following two setups:

#### Option 1: Manually Create a Kubernetes Secret

1. Log in to your Docker registry and follow the [Kubernetes documentation](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#log-in-to-docker-hub) to produce a `/.docker/config.json` file.

> 2. Reach out to [Astronomer Support](https://support.astronomer.io) and send the Namespace of the Deployment you want to use the KubernetesPodOperator with. A Deployment's namespace can be found in the Deployment view of the Astronomer UI.

Reviewer: ```suggestion
2. Reach out to [Astronomer Support](https://support.astronomer.io) and include the namespace of the Deployment you want to use the KubernetesPodOperator with. A Deployment's namespace can be found in the Deployment view of the Astronomer UI.
```

From here, Astronomer Support will give you instructions on how to securely send the output of your `/.docker/config.json` file. Do not send this file via email, as it contains sensitive credentials to your registry.

#### Option 2: Let Astronomer Create a Kubernetes Secret

Instead of creating a Kubernetes secret yourself, you can send Astronomer the following credentials:

- Astronomer Cluster
- Docker username
- Docker password
- Docker email
- Docker server (only if other than Dockerhub)

Astronomer Support will use this information to directly create a Kubernetes secret and `dockerconfigjson` file in the specified Cluster.

### Step 2: Run the KubernetesPodOperator

> Once Astronomer has added the Kubernetes secret to your Cluster, you will be notified and provided with the name of the secret.

Reviewer: ```suggestion
Once Astronomer has added the Kubernetes secret to your Deployment, you will be notified and provided with the name of the secret, represented as `<secret-name>` in the example below.
```

Author: Feels like we are combining two different parts of the setup in one sentence here. Let's keep the Astronomer support communication and the example usage separate. I update the name of the variable in the example to show that it's user supplied per our normal style - I think that's enough to communicate where to specify the secret.

- Execute custom Docker images at the task level, potentially with Python packages and dependencies that would otherwise conflict with the rest of your Deployment's dependencies
- If you know how much CPU and Memory your task consumes, you can specify those values as limits or minimums at the task level
- Flexibility to write task logic in a language other than Python
- Horizontal scaling that is cost-effective, dynamic, and minimally dependent on Worker resources
- Access to images hosted in private Docker repositories
- Kubernetes-native configuration, including the ability to set volumes, secrets, and affinities in the form of a YAML file

> - Coupled with the Data Plane, you have the ability to run Kubernetes pods without managing any underlying infrastructure

Author: I'm starting to question our convention of putting "This guide" statements at the end of overviews. Shouldn't we include these towards the beginning so that people know what they're getting into?

- `namespace = conf.get('kubernetes', 'NAMESPACE')`: Astronomer Cloud Deployments run on their own Kubernetes namespaces within a Cluster. The KubernetesPodOperator needs to know information about this namespace. Because the namespace variable is injected into your Deployment's `airflow.cfg` file, you can programmatically import this namespace using this setting.
- `image`: This is the image that the operator will use to run its defined task, commands, and arguments. The value you specify can be either an image tag that's publicly available on Dockerhub or a complete URL + image tag for another Docker registry. To pull an image from a private registry, read [Pull Images from a Private Registry](kubernetespodoperator.md#pull-images-from-a-private-registry).
- `in_cluster=True`: When this value is set, your task will look inside your Cluster for a Kubernetes configuration. This ensures that the workers running in the Pod have the correct privileges within the Cluster.
- `is_delete_operator_pod=True`: This setting ensures that the Pods of completed tasks are deleted, which lowers resource usage on your Cluster.

> Once you push this code to a Deployment, your Cluster will automatically spin up and down Pods to run tasks that use the KubernetesPodOperator.

Author: Some minor wording updates here.

## Configure Task-Level Pod Resources

> Astronomer Cloud automatically allocates resources to Pods created by the KubernetesPodOperator. Resources used by the KubernetesPodOperator are not limited, meaning that the operator could theoretically provision all your Cluster's available remaining CPU and memory to complete a task. Because of this, we recommend managing the KubernetesPodOperator's resource usage by specifying [compute resource requests and limits](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/) on a per-task basis.

Author: Not according to Kubernetes documentation: https://kubernetes.io/docs/concepts/workloads/pods/

> Astronomer Cloud automatically allocates resources to Pods created by the KubernetesPodOperator. Resources used by the KubernetesPodOperator are not limited, meaning that the operator could theoretically provision all your Cluster's available remaining CPU and memory to complete a task. Because of this, we recommend managing the KubernetesPodOperator's resource usage by specifying [compute resource requests and limits](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/) on a per-task basis.

Reviewer: Ok can u make consistent

To do so, define a dictionary of compute resources in your DAG. For example, applying the following dictionary to a Pod would ensure that the Pod runs with exactly 800m of CPU and 3Gi of memory at all times:

```python
compute_resources = \
  {'request_cpu': '800m',
  'request_memory': '3Gi',
  'limit_cpu': '800m',
  'limit_memory': '3Gi'}
```

To use a dictionary, specify the `resources` variable in your instantiation of the KubernetesPodOperator:

```python {11}
with dag:
    k = KubernetesPodOperator(
        namespace=namespace,
        image="hello-world",
        labels={"foo": "bar"},
        name="airflow-test-pod",
        task_id="task-one",
        in_cluster=in_cluster, # if set to true, will look in the cluster, if false, looks for file
        cluster_context='docker-for-desktop', # is ignored when in_cluster is set to True
        config_file=config_file,
        resources=compute_resources,
        is_delete_operator_pod=True,
        get_logs=True)
```

> ## Pull Images from a Private Registry

Author: Probably "Run" is safest? "Execute" implies that there's a set end state, like a task, whereas this is more like a platform for doing execution.

> ## Pull Images from a Private Registry

Reviewer: "Run" is cool with me

By default, the KubernetesPodOperator expects to pull an image that's hosted publicly on Dockerhub. If you want to pull images from a private registry, complete the setup in this topic.

### Prerequisites

To complete this setup, you need:

- A private registry that hosts your images.
- kubectl.
- An Astronomer Cloud project.

> - A Deployment hosting your project.

Author: Seems like a decent possibility of somebody searching for this setup and only following it without following anything above. In terms of header heirarchy, the prereqs for private image registry are not nested in the header for the general prereqs, so there's no structural connection there either.

### Step 1: Create a Kubernetes Secret

To securely use private Docker images on Astronomer Cloud, you need to create a Kubernetes secret for accessing your private Docker registry. To do this, complete one of the following two setups:

#### Option 1: Manually Create a Kubernetes Secret

> 1. Follow the [Kubernetes documentation](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#create-a-secret-by-providing-credentials-on-the-command-line) to create a Kubernetes secret for accessing your private registry. Note the name of the secret for the next step.

Reviewer: @jwitz Can you add this?

> 1. Follow the [Kubernetes documentation](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#create-a-secret-by-providing-credentials-on-the-command-line) to create a Kubernetes secret for accessing your private registry. Note the name of the secret for the next step.

Author: If we are already linking out to the k8s docs, then what is the benefit of including the snippet from the Kubernetes in our own docs? I'd rather have a user follow the k8s documentation completely rather than second guess that they have the right file by comparing the snippet in our docs to the external one.

2. Run the following command to confirm that the `dockerconfigjson` file was created correctly:

    ```sh
    kubectl get secret <secret-name> --output="jsonpath={.data.\.dockerconfigjson}" | base64 --decode
    ```

    Copy this output.

3. Send the output from Step 2 and the name of the Cluster that you are configuring to Astronomer Support. From here, Astronomer Support  will add your secret to the Cluster.

#### Option 2: Let Astronomer Create a Kubernetes Secret

Instead of creating a Kubernetes secret yourself, you can send Astronomer the following credentials:

- Astronomer Cluster
- Docker username
- Docker password
- Docker email
- Docker server (only if other than Dockerhub)

> Astronomer Support will use this information to directly create a Kubernetes secret and `dockerconfigjson` file in the specified Cluster.

Reviewer: @jwitz Kris is out for a bit and won't have time to confirm - let's just delete this Option #2 and add back this option later if #1 becomes a problem. Pretty sure we don't want to do this

### Step 2: Run the KubernetesPodOperator

Once Astronomer has added the Kubernetes secret to your Cluster, you will be notified and provided with the name of the secret.

> From here, you can use the KubernetesPodOperator with your private images by importing Kubernetes client Models and specifying `image_pull_secrets` in your operator instantiation:

Author: I kept the import step here since that's necessary for pulling the image
